Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/AbfsUri.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) Microsoft Corporation
+ *
+ * All rights reserved.
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+ * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+ * the Software.
+ *
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+ * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.microsoft.azure.hdinsight.common;
+
+import com.microsoft.azuretools.azurecommons.helpers.Nullable;
+import org.apache.commons.lang3.StringUtils;
+
+import java.net.URI;
+import java.util.UnknownFormatConversionException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class AbfsUri {
+    public static final String AdlsGen2PathPattern = "^(?<schema>abfss?)://(?<fileSystem>[^/.\\s]+)@(?<accountName>[^/.\\s]+)(\\.)(dfs\\.core\\.windows\\.net)(?<relativePath>(/[^\\s]+)*/?)$";
+    public static final String AdlsGen2RestfulPathPattern = "^(?<schema>https?)://(?<accountName>[^/.\\s]+)(\\.)(dfs\\.core\\.windows\\.net)(/)(?<fileSystem>[^/.\\s]+)(?<relativePath>(/[^\\s]+)*/?)$";
+    public static final Pattern ABFS_URI_PATTERN = Pattern.compile(AdlsGen2PathPattern, Pattern.CASE_INSENSITIVE);
+    public static final Pattern HTTP_URI_PATTERN = Pattern.compile(AdlsGen2RestfulPathPattern, Pattern.CASE_INSENSITIVE);
+
+    private final URI rawUri;
+    private final LaterInit<String> fileSystem = new LaterInit<>();
+    private final LaterInit<String> accountName = new LaterInit<>();
+    private final LaterInit<String> relativePath = new LaterInit<>();
```

Comment:
We may change `LaterInit<String>` to `LaterInit<Path>` later.

Reply:
If you are talking about `java.nio.file.Path`, there are some concerns here.
To initialize a `java.nio.file.Path`, we can use either ` Paths.get(String first, String... more)`or `Paths.get(URI uri)`. Please find the implementations below.
```
    public static Path get(String first, String... more) {
        return FileSystems.getDefault().getPath(first, more);
    }

    public static Path get(URI uri) {
        String scheme =  uri.getScheme();
        if (scheme == null)
            throw new IllegalArgumentException("Missing scheme");

        // check for default provider to avoid loading of installed providers
        if (scheme.equalsIgnoreCase("file"))
            return FileSystems.getDefault().provider().getPath(uri);

        // try to find provider
        for (FileSystemProvider provider: FileSystemProvider.installedProviders()) {
            if (provider.getScheme().equalsIgnoreCase(scheme)) {
                return provider.getPath(uri);
            }
        }

        throw new FileSystemNotFoundException("Provider \"" + scheme + "\" not installed");
    }

```
We can't use the second way to initialize a relative path since there are no "schema" defined for a relative path. Therefore, the only option is the first one.
But for the first option, the constructor has different behavior on different OS platforms. As far as I have tested, the following code has different output on Mac and Windows.
```
import java.nio.file.Path;
import java.nio.file.Paths;

public class TestPath {
    public static void testPathOnDifferentOs() {
        Path path = Paths.get("/sub/path/to");
        System.out.println(path.toString());
        System.out.println(path.toAbsolutePath().toString());
    }

    public static void main(String[] args) {
        testPathOnDifferentOs();
    }
}

```
On Windows, the output is
```
\sub\path\to
C:\sub\path\to
```

On Mac, the output is
```
/sub/path/to
/sub/path/to
```

Therefore, I prefer to use `String` here which is much more clean and flexible.

---
In reply to: [362136440](https://github.com/microsoft/azure-tools-for-java/pull/3885#discussion_r362136440) [](ancestors = 362136440)